### PR TITLE
[#1449] feat(catalogs): Introudce new module `bundled-catalog` for query engine.

### PR DESCRIPTION
**Pull request by yuqi1129**
### What changes were proposed in this pull request?

- Add a new module `bundled-catalog` to hold common information like property meta of all catalogs 
- Put `bundled-catalog` to `trino-connector` dependency and verify everything is OK. 

### Why are the changes needed?

We need a common module to hold the property meta of all catalogs so that the query engine depends on it can use it for query. 

Fix: #1449 

### Does this PR introduce _any_ user-facing change?

N/A

### How was this patch tested?

Add some UT in module `Trino-connector`

### CODE REVIEW

File: settings.gradle.kts [hunk 1]
```diff
@@ -14,3 +14,4 @@ include("clients:client-java", "clients:client-java-runtime")
 include("trino-connector")
 include("web")
 include("docs")
+include("catalogs:catalog-common")
```
Comment (diqiu50) [inline settings.gradle.kts:17]: Put it to line 12 is better.

File: catalogs/catalog-common/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  exclude("javax/**")
+  exclude("schema/**")
+  exclude("fr/**")
+  exclude("google/**")
+  exclude("groovy/**")
+  exclude("images/**")
+  exclude("**/*.conf")
+  exclude("**/*.so")
+  exclude("**/*.sxd")
+  exclude("**/*.ddl")
```
Comment (jerryshao) [inline catalogs/catalog-common/build.gradle.kts:53]: Did you exclude everything and only leave something you wanted?

File: catalogs/catalog-common/src/main/java/com/datastrato/catalog/common/ClassProvider.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2023 Datastrato Pvt Ltd.
```
Comment (jerryshao) [inline catalogs/catalog-common/src/main/java/com/datastrato/catalog/common/ClassProvider.java:2]: please update the license header.

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveSchemaPropertiesMetadata.java [hunk 4]
```diff
@@ -15,7 +15,7 @@
 
 public class HiveSchemaPropertiesMetadata extends BasePropertiesMetadata {
   public static final String LOCATION = "location";
-  private static final Map<String, PropertyEntry<?>> propertiesMetadata;
+  public static final Map<String, PropertyEntry<?>> propertiesMetadata;
```
Comment (jerryshao) [inline catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveSchemaPropertiesMetadata.java:18]: Why do we need to make this public?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/jdbc/mysql/MySQLDataTypeTransformer.java [hunk 5]
```diff
@@ -5,16 +5,24 @@
...
 /** Type transformer between MySQL and Trino */
 public class MySQLDataTypeTransformer extends GeneralDataTypeTransformer {
 
   @Override
   public Type getGravitinoType(io.trino.spi.type.Type type) {
+    for (Map.Entry<String, PropertyEntry<?>> entryEntry :
+        HiveTablePropertiesMetadata.propertiesMetadata.entrySet()) {
+      System.out.println(entryEntry.getKey() + " " + entryEntry.getValue());
+    }
+
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/jdbc/mysql/MySQLDataTypeTransformer.java]: Oh, it's the test code, I should remove it.

File: settings.gradle.kts (hunk 1 above)
Comment (yuqi1129) [reply settings.gradle.kts]: OK

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveSchemaPropertiesMetadata.java (hunk 4 above)
Comment (yuqi1129) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveSchemaPropertiesMetadata.java]: Mistake, I have reverted it.

File: catalogs/catalog-common/build.gradle.kts (hunk 2 above)
Comment (yuqi1129) [reply catalogs/catalog-common/build.gradle.kts]: Yes, the final jar is like this.

image

**Comment by yuqi1129** [comment]
@jerryshao @diqiu50 
Please take a look again, thanks.

File: catalogs/catalog-common/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+
+import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
```
Comment (jerryshao) [inline catalogs/catalog-common/build.gradle.kts:6]: I would suggest using a different package name, common will probably be used later on. you can think of a different name like "bundled-catalog" for example.

File: catalogs/catalog-common/src/main/java/com/datastrato/catalog/common/ClassProvider.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+import com.datastrato.gravitino.catalog.lakehouse.iceberg.IcebergSchemaPropertiesMetadata;
+import com.datastrato.gravitino.catalog.lakehouse.iceberg.IcebergTablePropertiesMetadata;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * {@link ClassProvider} holds the set of classes that are required for the query engine to use
+ * Gravitino, currently, it mainly focuses on the classes that are required for the property
+ * metadata.
+ */
```
Comment (jerryshao) [inline catalogs/catalog-common/src/main/java/com/datastrato/catalog/common/ClassProvider.java:24]: Can you add more comments about the reasons why we need to build this package.

File: trino-connector/src/test/java/com/datastrato/gravitino/trino/connector/catalog/hive/TestHiveCatalogPropertyConverter.java [hunk 8]
```diff
@@ -28,4 +30,13 @@ public void testConverter() {
     Assert.assertEquals(re.get("hive.compression-codec"), "ZSTD");
     Assert.assertEquals(re.get("hive.unknown-key"), null);
   }
+
+  @Test
+  // To test whether we can load property metadata from HiveTablePropertiesMetadata successfully.
+  public void testPropertyMetadata() {
+    for (Map.Entry<String, PropertyEntry<?>> entryEntry :
+        new HiveTablePropertiesMetadata().propertyEntries().entrySet()) {
+      System.out.println(entryEntry.getKey() + " " + entryEntry.getValue());
```
Comment (jerryshao) [inline trino-connector/src/test/java/com/datastrato/gravitino/trino/connector/catalog/hive/TestHiveCatalogPropertyConverter.java:39]: Why do we need this test? if you want to have a test, you should make it more formal. Also never use system output in the codes. As I remembered, it will fail the style check, but seems not.

Comment (yuqi1129) [reply trino-connector/src/test/java/com/datastrato/gravitino/trino/connector/catalog/hive/TestHiveCatalogPropertyConverter.java]: I want to test whether we can load `bundled-catalog` jar succfully, so, I use `Logger` instead?

File: catalogs/catalog-common/src/main/java/com/datastrato/catalog/common/ClassProvider.java (hunk 7 above)
Comment (yuqi1129) [reply catalogs/catalog-common/src/main/java/com/datastrato/catalog/common/ClassProvider.java]: Added.

File: trino-connector/src/test/java/com/datastrato/gravitino/trino/connector/catalog/hive/TestHiveCatalogPropertyConverter.java (hunk 8 above)
Comment (jerryshao) [reply trino-connector/src/test/java/com/datastrato/gravitino/trino/connector/catalog/hive/TestHiveCatalogPropertyConverter.java]: I think you should use junit api to verify, otherwise it is not a UT, IMO.

Comment (yuqi1129) [reply trino-connector/src/test/java/com/datastrato/gravitino/trino/connector/catalog/hive/TestHiveCatalogPropertyConverter.java]: OK, I will change it accordingly.

File: build.gradle.kts [hunk 9]
```diff
@@ -452,7 +452,7 @@ tasks {
     subprojects.forEach() {
       if (!it.name.startsWith("catalog") &&
         !it.name.startsWith("client") && it.name != "trino-connector" &&
-        it.name != "integration-test"
+        it.name != "integration-test" && !it.name.contains("bundled")
```
Comment (jerryshao) [inline build.gradle.kts:455]: Why using `contains`, not directly compare the strings?

File: catalogs/bundled-catalog/build.gradle.kts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  implementation(project(":core"))
+  implementation(project(":catalogs:catalog-hive"))
+  implementation(project(":catalogs:catalog-lakehouse-iceberg"))
+  implementation(project(":catalogs:catalog-jdbc-mysql"))
+  implementation(project(":catalogs:catalog-jdbc-postgresql"))
+}
+
+tasks.test {
+  useJUnitPlatform()
+}
```
Comment (jerryshao) [inline catalogs/bundled-catalog/build.gradle.kts:24]: Do you need this?

File: catalogs/catalog-hive/build.gradle.kts [hunk 11]
```diff
@@ -77,12 +77,12 @@ dependencies {
 tasks {
   val copyDepends by registering(Copy::class) {
     from(configurations.runtimeClasspath)
-    into("build/libs")
+    into("build/libs_all")
```
Comment (jerryshao) [inline catalogs/catalog-hive/build.gradle.kts:80]: What is the purpose of this change?

File: trino-connector/src/test/java/com/datastrato/gravitino/trino/connector/catalog/hive/TestHiveCatalogPropertyConverter.java [hunk 12]
```diff
@@ -5,13 +5,20 @@
...
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.testcontainers.shaded.com.google.common.collect.ImmutableMap;
 import org.testng.Assert;
 import org.testng.annotations.Test;
 
 public class TestHiveCatalogPropertyConverter {
 
+  private static final Logger LOG = LoggerFactory.getLogger(TestHiveCatalogPropertyConverter.class);
```
Comment (jerryshao) [inline trino-connector/src/test/java/com/datastrato/gravitino/trino/connector/catalog/hive/TestHiveCatalogPropertyConverter.java:20]: This is not used, seems can be removed.

File: catalogs/bundled-catalog/build.gradle.kts (hunk 10 above)
Comment (yuqi1129) [reply catalogs/bundled-catalog/build.gradle.kts:24]: I will remove it.

File: catalogs/catalog-hive/build.gradle.kts (hunk 11 above)
Comment (yuqi1129) [reply catalogs/catalog-hive/build.gradle.kts:80]: The `build` and `copyDepends` tasks of `catalog-hive` use the same output directory 'build/libs', which is needed by task `shadowJar` in `bundled-catalog`, gradle will get errors and says "task `shadowJar` depends on task `copyDepends`".

Comment (jerryshao) [reply catalogs/catalog-hive/build.gradle.kts:80]: You'd better add comment on the file to describe it.

Comment (yuqi1129) [reply catalogs/catalog-hive/build.gradle.kts:84]: Added.

**Review by jerryshao — APPROVED** [review]
